ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java

```diff
@@ -482,6 +507,100 @@ public void run() {
       }
     }
   }
+
+  /**
+   * A special {@link OperationController} that is responsible for handling background operations for this router.
+   *
+   * Background operations will be a scaling unit of its own (using its own {@link NetworkClient}), so that these
+   * operations do not interfere or contend with resources used for regular operations.
+   *
+   * Background operations include:
+   * 1. Deleting chunks of a composite blob that is deleted. When a composite blob is deleted, only the
+   * associated metadata blob is deleted before notifying the caller. This keeps the latency low. In the background,
+   * the associated metadata blob will be fetched, the chunk ids will be extracted and deleted.
+   *
+   * 2. (TBD) Deleting successfully put chunks of a failed composite blob put operation. Today, this is done by the
+   * same {@link OperationController} doing the put.
+   */
+  private class BackgroundDeleter extends OperationController {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Instantiate the BackgroundDeleter
+     * @throws IOException if the associated {@link OperationController} throws one.
+     */
+    BackgroundDeleter() throws IOException {
+      super("backgroundDeleter");
+      super.putManager.close();
+    }
+
+    /**
+     * Put operations are disallowed in the BackgroundDeleter.
+     */
+    @Override
+    protected void putBlob(BlobProperties blobProperties, byte[] userMetadata, ReadableStreamChannel channel,
+        FutureResult<String> futureResult, Callback<String> callback) {
+      RouterException routerException =
+          new RouterException("Illegal attempt to put blob through backgroundDeleteOperationController",
+              RouterErrorCode.UnexpectedInternalError);
+      routerMetrics.operationDequeuingRate.mark();
+      routerMetrics.onPutBlobError(routerException);
+      operationCompleteCallback.completeOperation(futureResult, callback, null, routerException);
+    }
+
+    /**
+     * Initiate the deletes of the data chunks associated with this blobId, if this blob turns out to be a composite
+     * blob.
+     * @param blobId the blobId string associated with the possibly composite blob.
+     */
+    void initiateChunkDeletesIfAny(String blobId) {
+      FutureResult<List<StoreKey>> futureResult = new FutureResult<>();
+      Callback<List<StoreKey>> callback = new Callback<List<StoreKey>>() {
+        @Override
+        public void onCompletion(List<StoreKey> result, Exception exception) {
+          if (exception != null) {
+            logger.error("Encountered exception when attempting to get chunks of a possibly composite deleted blob",
+                exception);
+          } else if (result != null) {
+            // result can be null if the blob is not a composite blob.
+            for (StoreKey key : result) {
+              currentOperationsCount.incrementAndGet(); // @todo
```

Comment:
Not related to this pr. I found `currentOperationsCount` is named a bit confusing. My first understanding is that it is a gauge-like metric that records the current operations, but it seems a so-far operation metric.